lib/phoenix/tracker/state.ex

```diff
@@ -196,17 +198,24 @@ defmodule Phoenix.Tracker.State do
   end
   def extract(%State{mode: :normal, values: values, clouds: clouds} = state, remote_ref, remote_context) do
     pruned_clouds = Map.take(clouds, Map.keys(remote_context))
-    map = foldl(values, [], fn
-      {{_topic, _pid, _key}, _meta, {^remote_ref, _clock}}, acc -> acc
-      {{topic, pid, key}, meta, {replica, _clock} = tag}, acc ->
-        if Map.has_key?(remote_context, replica) do
-          [{tag, {pid, topic, key, meta}} | acc]
+    # fn {{topic, pid, key}, meta, {replica, clock}} when replica !== remote_ref ->
+    #  {{replica, clock}, {pid, topic, key, meta}}
+    # end
+    ms = [{
+      {{:"$1", :"$2", :"$3"}, :"$4", {:"$5", :"$6"}},
+      [{:"=/=", :"$5", {:const, remote_ref}}],
+      [{{{{:"$5", :"$6"}}, {{:"$2", :"$1", :"$3", :"$4"}}}}]
+    }]
+    data =
+      foldl(values, [], ms, fn {{replica, _} = tag, data}, acc ->
+        if match?(%{^replica => _}, remote_context) do
+          [{tag, data} | acc]
         else
           acc
         end
-    end) |> :maps.from_list()
```

Comment:
@josevalim @chrismccord there is something weird going on in that function even before extracting.

When called in tests `remote_ref` is an atom, but the `replica` value is a pair `{atom, integer}` so it can never match - keys in the `remote_conetxt` map have a similar format.

This makes me wonder what is broken - the implementation that assumes `replica` and `remote_ref` might match or tests.

Reply:
I pushed a commit that makes correct use of `State.extract` in tests. I'm not sure why the tests did not fail previously and if we're missing some test to check this particular behaviour.